From dev-bounces@openvswitch.org Fri Apr 16 01:12:35 2010
Received: from FTLPMAILMX02.citrite.net (10.9.154.224) by
 LONPMAILMX02.citrite.net (10.30.224.163) with Microsoft SMTP Server (TLS)
 id 8.1.393.1; Fri, 16 Apr 2010 01:12:35 +0100
Received: from SMTP02.CITRIX.COM (10.9.154.239) by FTLPMAILMX02.citrite.net
 (10.9.154.224) with Microsoft SMTP Server id 8.1.393.1; Thu, 15 Apr 2010
 20:12:33 -0400
Received: from vps.nicira.com ([66.7.219.28])  by SMTP02.CITRIX.COM with
 ESMTP/TLS/DHE-RSA-AES256-SHA; 15 Apr 2010 20:12:32 -0400
Received: from localhost ([127.0.0.1]:52682 helo=vps.nicira.com)	by
 vps.nicira.com with esmtp (Exim 4.69)	(envelope-from
 <dev-bounces@openvswitch.org>)	id 1O2ZAC-0002zQ-L3; Thu, 15 Apr 2010
 20:12:04 -0400
Received: from 173-164-153-21-sfba.hfc.comcastbusiness.net
 ([173.164.153.21]:34978 helo=leadville.nicira.com)	by vps.nicira.com with
 esmtps (TLSv1:AES256-SHA:256) (Exim 4.69)	(envelope-from <blp@nicira.com>)
 id 1O2ZA9-0002xP-Ub	for dev@openvswitch.org; Thu, 15 Apr 2010 20:12:02 -0400
Received: from hardrock.nicira.com ([172.16.0.20])	by leadville.nicira.com
 with esmtp (Exim 4.67)	(envelope-from <blp@nicira.com>)	id
 1O2ZA4-0007f4-OM; Thu, 15 Apr 2010 17:11:56 -0700
Received: from blp by hardrock.nicira.com with local (Exim 4.67)
 (envelope-from <blp@hardrock.nicira.com>)	id 1O2ZAF-0006DN-Px; Thu, 15 Apr
 2010 17:12:07 -0700
From: Ben Pfaff <blp@nicira.com>
To: "dev@openvswitch.org" <dev@openvswitch.org>
CC: Ben Pfaff <blp@nicira.com>, Henrik Amren <henrik@nicira.com>
Sender: "dev-bounces@openvswitch.org" <dev-bounces@openvswitch.org>
Date: Fri, 16 Apr 2010 01:11:59 +0100
Subject: [ovs-dev] [PATCH 4/4] bridge: Don't learn from inadmissible flows
 when revising learning table.
Thread-Topic: [ovs-dev] [PATCH 4/4] bridge: Don't learn from inadmissible
 flows	when revising learning table.
Thread-Index: Acrc+YP5Ti7B7fagT7u0OR1qIjKQbA==
Message-ID: <1271376719-23848-4-git-send-email-blp@nicira.com>
References: <1271376719-23848-1-git-send-email-blp@nicira.com>
List-Help: <mailto:dev-request@openvswitch.org?subject=help>
List-Subscribe:
 <http://openvswitch.org/mailman/listinfo/dev_openvswitch.org>,
 <mailto:dev-request@openvswitch.org?subject=subscribe>
List-Unsubscribe:
 <http://openvswitch.org/mailman/options/dev_openvswitch.org>,
 <mailto:dev-request@openvswitch.org?subject=unsubscribe>
In-Reply-To: <1271376719-23848-1-git-send-email-blp@nicira.com>
Accept-Language: en-GB, en-US
Content-Language: en-US
X-MS-Exchange-Organization-AuthAs: Anonymous
X-MS-Exchange-Organization-AuthSource: FTLPMAILMX02.citrite.net
X-MS-Has-Attach: 
X-Auto-Response-Suppress: All
X-MS-TNEF-Correlator: 
x-sbrs: 5.3
x-mesageid: 92502212
x-ironport-server: ftlpip02.citrite.net
x-remote-ip: 66.7.219.28
x-policy: $ACCEPTED
x-ironport-anti-spam-filtered: true
x-ironport-anti-spam-result:
 AqABAC9Ix0tCB9scmWdsb2JhbACbaxUBAQEBAQgLCgcRIr5Qgk+CPwSDKw
x-ironport-av: E=Sophos;i="4.52,215,1270440000";   
 d="scan'208";a="92502212"
x-antiabuse: Sender Address Domain - openvswitch.org
x-source: 
x-source-args: 
x-source-dir: 
list-id: <dev_openvswitch.org.openvswitch.org>
errors-to: dev-bounces@openvswitch.org
list-post: <mailto:dev@openvswitch.org>
list-archive: <http://openvswitch.org/pipermail/dev_openvswitch.org>
x-beenthere: dev@openvswitch.org
x-mailman-version: 2.1.12.cp3
Content-Type: text/plain; charset="us-ascii"
MIME-Version: 1.0
X-Evolution-Source: imap://ianca@lonpmailmx01.citrite.net/
Content-Transfer-Encoding: 8bit

Various kinds of flows are inadmissible and must be dropped.  Most notably,
OVS drops packets received on a bond whose destinations are ones that OVS
has already learned on a different port.  As the comment says:

         /* Drop all packets for which we have learned a different input
          * port, because we probably sent the packet on one slave and got
          * it back on the other.  Broadcast ARP replies are an exception
          * to this rule: the host has moved to another switch. */

As an important side effect of dropping these packets, OVS does not use
them for MAC learning when it sets up the corresponding flows.

However, OVS also periodically scans the datapath flow table and uses
information about flow activity to update its learning tables.  (Otherwise,
learning table entries could expire because no new flows were being set up,
even though active flows existed.)  This process, implemented in
bridge_account_flow_ofhook_cb(), did not check for admissibility, so
packets received on a bond could be used for learning even though another
port had already been learned.

This commit fixes the problem by making bridge_account_flow_ofhook_cb()
check for admissibility.

QA notes: Reproducing this problem requires some care and some luck.  One
way is to have two VMs with network interfaces on a single bonded network.
Both bonded interfaces must be up (otherwise packets sent out on one slave
will never be received on the other).  The problem will also not occur if
the physical switch that the bond slaves are plugged into has learned the
MAC address of the VMs involved (because the physical switch will then,
again, drop the packets without sending them back in on the other slave).
Finally, there needs to be some luck in timing and perhaps with the OVS
internal hash function also.

(One way to reproduce it reliably is to plug a pair of Ethernet ports into
each other with a cable, without an intermediate switch, and then use that
pair of ports as a bond.  Then every packet sent out on one will
immediately be received on the other, triggering the problem fairly often.
If this doesn't work at first, try changing the Ethernet address used on
one side or the other.)

To verify that the problem being observed is the one fixed by this commit,
turn on bridge debugging with "ovs-appctl vlog/set bridge:file" and look
for "bridge xapi2: learned that 00:01:02:03:04:06 is on port bond0 in VLAN
0" where 00:01:02:03:04:06 is a VM's Ethernet address and bond0 is the
name of the bond in the ovs-vswitchd log file.

Testing: I ran the "loopback bond" test above with and without this commit,
twice each in case I was just lucky.

CC: Henrik Amren <henrik@nicira.com>

Bug #2366.
Bug NIC-64.
Bug NIC-69.
---
 vswitchd/bridge.c |   12 +++++-------
 1 files changed, 5 insertions(+), 7 deletions(-)

diff --git a/vswitchd/bridge.c b/vswitchd/bridge.c
index d33944a..2d60525 100644
--- a/vswitchd/bridge.c
+++ b/vswitchd/bridge.c
@@ -2380,18 +2380,16 @@ bridge_account_flow_ofhook_cb(const flow_t *flow,
                               void *br_)
 {
     struct bridge *br = br_;
-    struct port *in_port;
     const union odp_action *a;
+    struct port *in_port;
+    tag_type tags = 0;
+    int vlan;
 
     /* Feed information from the active flows back into the learning table
      * to ensure that table is always in sync with what is actually flowing
      * through the datapath. */
-    in_port = port_from_dp_ifidx(br, flow->in_port);
-    if (in_port) {
-        int vlan = flow_get_vlan(br, flow, in_port, false);
-         if (vlan >= 0) {
-            update_learning_table(br, flow, vlan, in_port);
-        }
+    if (is_admissible(br, flow, false, &tags, &vlan, &in_port)) {
+        update_learning_table(br, flow, vlan, in_port);
     }
 
     if (!br->has_bonded_ports) {
-- 
1.6.6.1


_______________________________________________
dev mailing list
dev@openvswitch.org
http://openvswitch.org/mailman/listinfo/dev_openvswitch.org

